core/indexer/process/transactions/grouper.go

```diff
@@ -0,0 +1,191 @@
+package transactions
+
+import (
+	"github.com/ElrondNetwork/elrond-go/core/indexer/types"
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/data/block"
+	"github.com/ElrondNetwork/elrond-go/data/receipt"
+	"github.com/ElrondNetwork/elrond-go/data/rewardTx"
+	"github.com/ElrondNetwork/elrond-go/data/smartContractResult"
+	"github.com/ElrondNetwork/elrond-go/data/transaction"
+)
+
+type txGrouper struct {
+	selfShardID    uint32
+	txBuilder      *txDBBuilder
+	calculateHash  func(object interface{}) ([]byte, error)
+	isInImportMode bool
+}
+
+func newTxGrouper(
+	txBuilder *txDBBuilder,
+	calculateHash func(object interface{}) ([]byte, error),
+	isInImportMode bool,
+	selfShardID uint32,
+) *txGrouper {
+	return &txGrouper{
+		txBuilder:      txBuilder,
+		calculateHash:  calculateHash,
+		selfShardID:    selfShardID,
+		isInImportMode: isInImportMode,
+	}
+}
+
+func (tg *txGrouper) groupNormalTxsAndRewards(
+	body *block.Body,
+	txPool map[string]data.TransactionHandler,
+	header data.HeaderHandler,
+	alteredAddresses map[string]*types.AlteredAccount,
+) (
+	map[string]*types.Transaction,
+	[]*types.Transaction,
+) {
+	transactions := make(map[string]*types.Transaction)
+	rewardsTxs := make([]*types.Transaction, 0)
+
+	for _, mb := range body.MiniBlocks {
+		mbHash, err := tg.calculateHash(mb)
+		if err != nil {
+			continue
+		}
+
+		mbStatus := computeStatus(tg.selfShardID, mb.ReceiverShardID)
+
+		switch mb.Type {
+		case block.TxBlock:
+			txs := getTransactions(txPool, mb.TxHashes)
+			for hash, tx := range txs {
```

Comment:
nested for in switch that is nested in another for. We should stop writing this code. Is ugly.

Reply:
not done

Reply:
refactored